indexing-service/src/main/java/org/apache/druid/indexing/common/SegmentLock.java

```diff
@@ -0,0 +1,237 @@
+
+  private final TaskLockType lockType;
+  private final String groupId;
+  private final String dataSource;
+  private final Interval interval;
+  private final String version;
+  private final int partitionId;
+  private final int priority;
+  private final boolean revoked;
+
+  @JsonCreator
+  public SegmentLock(
+      @JsonProperty("lockType") TaskLockType lockType,
+      @JsonProperty("groupId") String groupId,
+      @JsonProperty("dataSource") String dataSource,
+      @JsonProperty("interval") Interval interval,
+      @JsonProperty("version") String version,
+      @JsonProperty("partitionId") int partitionId,
+      @JsonProperty("priority") int priority,
+      @JsonProperty("revoked") boolean revoked
+  )
+  {
+    this.lockType = Preconditions.checkNotNull(lockType, "lockType");
+    this.groupId = Preconditions.checkNotNull(groupId, "groupId");
+    this.dataSource = Preconditions.checkNotNull(dataSource, "dataSource");
+    this.interval = Preconditions.checkNotNull(interval, "interval");
+    this.version = Preconditions.checkNotNull(version, "version");
+    this.partitionId = partitionId;
+    this.priority = priority;
+    this.revoked = revoked;
+  }
+
+  public SegmentLock(
+      TaskLockType lockType,
+      String groupId,
+      String dataSource,
+      Interval interval,
+      String version,
+      int partitionId,
+      int priority
+  )
+  {
+    this(lockType, groupId, dataSource, interval, version, partitionId, priority, false);
+  }
+
+  @JsonProperty
+  @Override
+  public String getType()
+  {
+    return TYPE;
+  }
+
+  @Override
+  public TaskLock revokedCopy()
+  {
+    return new SegmentLock(lockType, groupId, dataSource, interval, version, partitionId, priority, true);
+  }
+
+  @Override
+  public TaskLock withPriority(int newPriority)
+  {
+    return new SegmentLock(lockType, groupId, dataSource, interval, version, partitionId, newPriority, revoked);
+  }
+
+  @Override
+  public LockGranularity getGranularity()
+  {
+    return LockGranularity.SEGMENT;
+  }
+
+  @JsonProperty
+  @Override
+  public TaskLockType getLockType()
+  {
+    return lockType;
+  }
+
+  @JsonProperty
+  @Override
+  public String getGroupId()
+  {
+    return groupId;
+  }
+
+  @JsonProperty
+  @Override
+  public String getDataSource()
+  {
+    return dataSource;
+  }
+
+  @JsonProperty
+  @Override
+  public Interval getInterval()
+  {
+    return interval;
+  }
+
+  @JsonProperty
+  public int getPartitionId()
+  {
+    return partitionId;
+  }
+
+  @JsonProperty
+  @Override
+  public String getVersion()
+  {
+    return version;
+  }
+
+  @JsonProperty
+  @Override
+  public Integer getPriority()
+  {
+    return priority;
+  }
+
+  @Override
+  public int getNonNullPriority()
+  {
+    return priority;
+  }
+
+  @JsonProperty
+  @Override
+  public boolean isRevoked()
+  {
+    return revoked;
+  }
+
+  @Override
+  public boolean conflict(LockRequest request)
+  {
+    if (request instanceof TimeChunkLockRequest) {
```

Comment:
I think this method would be more readable if it checked for `instanceof LockRequestForNewSegment` first and then `!dataSource.equals(request.getDataSource()` as separate checks that return false (i.e., the cases that will never conflict), before going to the `TimeChunkLockRequest` and `SpecificSegmentLockRequest` blocks

Reply:
Done.